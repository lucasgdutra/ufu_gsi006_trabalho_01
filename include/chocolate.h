/* --------------------------------------------------------------------------------------
   Nesse arquivo temos o header da parte do chocolate, haja vista que teremos uma pilha
   de chocolate que será distruibida como brinde no nosso restaurante. Portanto, é preciso 
   que eu consiga ter uma pilha de chocolate e aloca-los nessa estrutura.
   Objetivo: Trabalho da disciplina Estrutura de dados (GSI006) BSI - Facom -
UFU Restricoes:

   Autor: Dalmo Scalon Inacio/ Lucas Gabriel Dutra de Souza
   Disciplina: Estrutura de Dados 1
   Data da ultima atualizacao: 04/01/2023
----------------------------------------------------------------------------------------*/
#ifndef CHOCOLATE_H
#define CHOCOLATE_H

#include "./pilha.h"

/*
	Struct chamada Chocolate, que contem um ponteiro pra char indicando nome e
   um int indicando tamanho desse nome.
*/
typedef struct chocolate {
	char *nome;
	int tamanho_nome;
} Chocolate;

/*
 Nome: inicializa_chocolate
 Entrada: Ponteiro do tipo Chocolate, ponteiro do tipo char indicando nome, e
 int tam_nome Processo:  Valida os ponteiros recebidos, e atribui nome e
 tamanho_nome para a struct Chocolate recebida
 Saida: return 0 caso dê certo, e 1 caso dê errado.
 */
int inicializa_chocolate(Chocolate *chocolate, char *nome, int tamanho_nome);

/*
	Struct Pilha_Chocolate que contem um ponteiro do tipo Pilha, para
   possibilitar criarmos nossa Pilha de Chocolate.
*/
typedef struct pilha_chocolate {
	Pilha *pilha;
} Pilha_Chocolate;

/*
 Nome: inicializa_pilha_chocolate
 Entrada: A função recebe um ponteiro do tipo Pilha_Chocolate.
 Processo: Função verifica se os ponteiros sao validos, posteriormente aloca a
 Pilha e passa a pilha alocada para pc. 
 Saida: return 0 caso dê certo, e 1 caso dê errado.
 */
int inicializa_pilha_chocolate(Pilha_Chocolate *pilha_chocolate);

/*
 Nome: empilha_chocolate
 Entrada: A função recebe um ponteiro do tipo Pilha_Chocolate e um ponteiro do
 tipo Chocolate Processo: Função verifica se os ponteiros sao validos, empilha o
 chocolate que é passado pra função. 
 Saida: return 0 caso dê certo, e 1 caso dê errado.
 */
int empilha_chocolate(Pilha_Chocolate *pilha_chocolate, Chocolate *chocolate);

/*
 Nome: desempilha_chocolate
 Entrada: A função recebe um ponteiro do tipo Pilha_Chocolate e um ponteiro do
 tipo Chocolate Processo: Função verifica se os ponteiros sao validos,
 desempilha o chocolate caso seja possivel e passa o ponteiro choco_retorno para
 caso desempilhar, receber o elemento desempilhado.
 Saida: return 0 caso dê certo, e 1 caso dê errado.
 */
int desempilha_chocolate(Pilha_Chocolate *pilha_chocolate,
						 Chocolate *retorno_chocolate);

#endif
