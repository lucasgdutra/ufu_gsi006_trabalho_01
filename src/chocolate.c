#include "../include/chocolate.h"
#include "../include/pilha.h"
#include <stdio.h>
#include <stdlib.h>

int inicializa_chocolate(Chocolate *c, char *nome, int tamanho_nome) {
    c->nome = nome;
    c->tamanho_nome = tamanho_nome;
    return 0;
};

int inicializa_pilha_chocolate(Pilha_Chocolate *pc){
    Pilha p;
    inicializa_pilha(&p, sizeof(Chocolate));

    pc->pilha = &p;
    return 0;
};


int empilha_chocolate(Pilha_Chocolate *pc, Chocolate *c){
    empilha(pc->pilha,c);
    return 0;
}

int desempilha_chocolate(Pilha_Chocolate *pc, Chocolate *choco_retorno){
    desempilha(pc->pilha,choco_retorno);
    return 0;
}