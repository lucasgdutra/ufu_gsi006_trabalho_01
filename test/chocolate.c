#include "../include/chocolate.h"
#include <stdio.h>
#include <stdlib.h>
#define MAX_CHAR_SIZE 30
int main() {
	char nome[MAX_CHAR_SIZE];
	while (scanf("%s", nome) != EOF) {
		Chocolate chocolate;
		inicializa_chocolate(&chocolate, nome, MAX_CHAR_SIZE);
		printf("nome: %s\n", chocolate.nome);
	}

	return 0;
}