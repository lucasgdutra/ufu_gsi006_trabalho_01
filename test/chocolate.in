laka
chokito
diamante_negro
alpino
batom
ferrero_rocher
prestigio