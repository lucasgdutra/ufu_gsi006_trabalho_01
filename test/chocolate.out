nome: laka
nome: chokito
nome: diamante_negro
nome: alpino
nome: batom
nome: ferrero_rocher
nome: prestigio
